### ISSUE TITLE
With defaultContentLanguageInSubdir: true, a 404.html file should be created at the root

### ISSUE THREAD

**Issue by divinerites** [gohugoio/hugo#5161]
In the direct relation with #5160  I open this (related ?) issue.

When option defaultContentLanguageInSubdiris set to true, a sitemap.xml file is automagically generated by Hugo at the root of the website, as well as in all language subdirectories.

I propose that the same is done for the 404.html file, which is currently (as of Hugo 0.48) only generated in language subdirectories but not at the root of the website.

See : 

https://discourse.gohugo.io/t/404-html-not-generated-on-root-if-defaultcontentlanguageinsubdir-true/14048

**Comment by martignoni** [comment]
Still present in Hugo v0.53.

**Comment by vitaliytv** [comment]
temporary workaround:
`NODE_ENV=production hugo -v && cp public/ru/404.html public/404.htm`

**Comment by yktoo** [comment]
I believe this issue is absolutely relevant. You need to use some nasty hacks otherwise.

**Comment by guillaumemaiano** [comment]
Would be nice to know if the "hack" is currently the recommended method to use.

**Comment by Lynxiayel** [comment]
I think this is not a simple problem that can be addressed with a cp hack. refer to #3075.
Some multilingual sites may want to have 404 page in different language. This involves in-site routing, and is not solvable with a single-languaged global 404.

For example, if I have a site with all English content under /en/ and their Chinese counterparts under /zh/. 
If a user accesed sth. not existed under /zh/, we want to present a Chinese 404 page, and in that page, all links (eg. home/section navs) should all be pointed still to pages under /zh/. The same goes for English pages. Apparently, in this case, showing either an English or Chinese 404 page may mess up the in-site routing. 

I personally think this is something need to be addressed by the server, and hugo (as a **static** site generator) can hardly do anything about it.

**Comment by onedrawingperday** [comment]
I became aware of this issue today.

Actually there is a much simpler way than what was posted above.

Generate the project locally and then simply copy the 404.html from the language directory under /static/.

The `staticDir` by default lives outside Hugo’s Multilingual Mode, therefore a 404.html under the root of `/static/` will be served under the root of the domain without having to use any shell command during deployment.

**Comment by Guite** [comment]
> Generate the project locally and then simply copy the 404.html from the language directory under /static/.

I guess this will lead to problems as soon as asset URLs will change? You would need to update your static `404.html` copy each time the site is rebuilt to ensure it still references valid assets.

**Comment by onedrawingperday** [comment]
Fair enough @Guite 

I tested further and today one can do the following in a multilingual project:

```
[module]
  [[module.mounts]]
    source = "content/404.md"
    target = "public/404.html"
```
The above will output a 404.html at the root of the project even with `defaultContentLanguageInSubdir = true`

I believe that this is a non-issue and could be closed.

cc: @jmooring @bep @divinerites

**Comment by onedrawingperday** [comment]
Fair enough @Guite 

I tested further and found that one can simply set the `url` parameter with a forward slash at the beginning, so that Hugo outputs the page under the host root, even if the project uses `defaultContentLanguageInSubdir = true`

e.g. `url = "/404.html`

I believe that this is a non-issue that can be closed since it is dead easy to output the 404 page at the root.

cc: @jmooring @divinerites @bep

**Comment by Guite** [comment]
@onedrawingperday  where do you define this `url` parameter`? Does this require creating the `404` page within `content/` explicitly? please elaborate.

**Comment by onedrawingperday** [comment]
@Guite 

Yes. Simply create a 404.md and include the above in its front matter.

**Comment by Guite** [comment]
Nice solution, thank you.

**Comment by tomy0000000** [comment]
@bep Can you look into this strange behavior? This has been stale for quite some time but could be fixed in a short time. A similar case worth mentioning is #5160, but for `robots.txt`. I'm unfamiliar with golang, so I'm not sure how to work on this problem, but if you can give some direction and hints, I can help work on this bug.

**Comment by anasfanani** [comment]
For 404 in multilingual site. I have this idea 
1. Create a route with cloudflare
2. I must save user lang setting in localStorage, then create a custom 404 in static dir for redirect to /en/404.html or id/404.html

**Comment by freestingo** [comment]
just successfully managed to get a custom multilanguage 404 page to show up in my Vercel deployment with this setup:
- `themes/<theme-name>/layouts/404.html`
  - all strings inside this template are translated with the `T` function (example: `{{ "pageNotFound" | T }}`)
  - in my case, I have both `en` and `it` languages set up, so when deploying this template will generate both `public/en/404.html` and `public/it/404.html`
- `vercel.json`
  - contents:
  ```json
  {
    "routes": [
      { "handle": "filesystem" },
      { "src": "/en/(.*)", "status": 404, "dest": "/en/404.html" },
      { "src": "/it/(.*)", "status": 404, "dest": "/it/404.html" },
      { "src": "/(.*)", "status": 404, "dest": "/en/404.html" }
    ]
  }
  ```
  - with this file I can [fine-tune](https://vercel.com/guides/custom-404-page#what-if-i-need-to-name-my-404-file-something-different) what translation the user sees when navigating to an unknown page (i.e. `/en/unknown-page` will show the English 404 page, while `/it/unknown-page` will show the Italian one; the English one is used as fallback for all URLs that do not start with any language code)

**Comment by CyrusYip** [comment]
I summarized three workarounds to this issue in a blog post: [Create Top-Level 404 Page for a Multilingual Hugo Site - Cyrus Yip's Blog](https://cyrusyip.org/en/post/2023/11/06/hugo-top-level-404/). I hope it will be helpful.
